### PR TITLE
docs: initial user mode documentation

### PR DESCRIPTION
**Pull request by tyrantlizard77**
This initial batch of documentation describes policies and
mechanism related to kernel objects and system calls.

Some details on porting user mode to a new arch have been
provided in the architecture porting guide.

Thread documentation updated with some user mode consideration.

This is not the final documentation, more to come in subsequent
patches.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

### CODE REVIEW

File: doc/kernel/usermode/kernelobjects.rst
Comment (lpereira) [inline doc/kernel/usermode/kernelobjects.rst]: Typo here, should end in `,` instead of `;`

File: doc/kernel/usermode/kernelobjects.rst
Comment (lpereira) [inline doc/kernel/usermode/kernelobjects.rst]: Missing `;`

File: doc/porting/arch.rst
Comment (lpereira) [inline doc/porting/arch.rst]: This is kind of obvious from the description below, but might be a good idea to note that this alone isn't going to stop any bullets.

**Review by lpereira — APPROVED** [review]
Seems pretty good overall.  Found just a few typos.  I'd just wait for @dbkinder for the real approval stamp.

File: doc/porting/arch.rst
Comment (andrewboie) [reply doc/porting/arch.rst]: done

File: doc/kernel/threads/lifecycle.rst
Comment (ramakrishnapallala) [inline doc/kernel/threads/lifecycle.rst]: if thread stack is cleared then will it restart from therad_entry()?

Comment (andrewboie) [reply doc/kernel/threads/lifecycle.rst]: yes, correct

Comment (andrewboie) [reply doc/kernel/threads/lifecycle.rst]: if you look at the API for k_thread_user_mode_enter() it takes an entry point and 3 parameters

**Review by ramakrishnapallala — APPROVED** [review]

File: doc/kernel/usermode/kernelobjects.rst
Comment (dbkinder) [inline doc/kernel/usermode/kernelobjects.rst]: needed for creating

File: doc/kernel/usermode/kernelobjects.rst
Comment (dbkinder) [inline doc/kernel/usermode/kernelobjects.rst]: nice to mark scripts with double back-ticks:

     ``scripts/gen_kobject_list.py``

File: doc/kernel/usermode/kernelobjects.rst
Comment (dbkinder) [inline doc/kernel/usermode/kernelobjects.rst]: An example would clarify this.

File: doc/kernel/usermode/kernelobjects.rst
Comment (dbkinder) [inline doc/kernel/usermode/kernelobjects.rst]: nice to mark scripts with double back-ticks:

     ``scripts/gen_kobject_list.py``

File: doc/kernel/usermode/kernelobjects.rst
Comment (dbkinder) [inline doc/kernel/usermode/kernelobjects.rst]: again, an example would help clarify this

File: doc/kernel/usermode/syscalls.rst
Comment (dbkinder) [inline doc/kernel/usermode/syscalls.rst]: How about:

     User threads run with a reduced set of privileges than Supervisor threads: certain
     CPU instructions may not be used, and they have access to only a limited part
     of the memory map.   System calls (may) allow user threads to perform operations
     not directly available to them.

File: doc/kernel/usermode/syscalls.rst
Comment (dbkinder) [inline doc/kernel/usermode/syscalls.rst]: mark filename with double back-ticks:

     under ``include/``

File: doc/kernel/usermode/syscalls.rst
Comment (dbkinder) [inline doc/kernel/usermode/syscalls.rst]: use the double back-ticks on all references to gen_syscalls.py

File: doc/kernel/usermode/syscalls.rst
Comment (dbkinder) [inline doc/kernel/usermode/syscalls.rst]: delete "and cannot be preceded by anything else" (that's what "be the first thing" means)

File: doc/kernel/usermode/syscalls.rst
Comment (dbkinder) [inline doc/kernel/usermode/syscalls.rst]: should the "special generated header" be more properly called a "special generated include"?

File: doc/kernel/usermode/syscalls.rst
Comment (dbkinder) [inline doc/kernel/usermode/syscalls.rst]: use double back-ticks on the file name here

File: doc/kernel/usermode/syscalls.rst
Comment (dbkinder) [inline doc/kernel/usermode/syscalls.rst]: convention is to put the double colon at the end of the sentence to start a code block:

     ``include/kernel.h``. At the bottom of ``include/kernel.h`` is::

        #include <syscalls/kernel.h>

File: doc/kernel/usermode/syscalls.rst
Comment (dbkinder) [inline doc/kernel/usermode/syscalls.rst]: change "which" to "that" here

File: doc/kernel/usermode/syscalls.rst
Comment (dbkinder) [inline doc/kernel/usermode/syscalls.rst]: use single back-ticks here:

     :c:func:`k_sem_init()`

File: doc/kernel/usermode/syscalls.rst
Comment (dbkinder) [inline doc/kernel/usermode/syscalls.rst]: This seems contradictory. Either you validate parameters or you don't.  Maybe just say validation of arguments (with assertions) is optional.

File: doc/kernel/usermode/syscalls.rst
Comment (dbkinder) [inline doc/kernel/usermode/syscalls.rst]: How about:

     All implementation functions must follow the naming convention, 
     name of the API prefixed with ``_impl_``.

File: doc/kernel/usermode/syscalls.rst
Comment (dbkinder) [inline doc/kernel/usermode/syscalls.rst]: delete "is when"

File: doc/kernel/usermode/syscalls.rst
Comment (dbkinder) [inline doc/kernel/usermode/syscalls.rst]: Better would be:

     Handler functions should be declared using these macros.

File: doc/kernel/usermode/syscalls.rst
Comment (dbkinder) [inline doc/kernel/usermode/syscalls.rst]: delete "which are used"

File: doc/kernel/usermode/syscalls.rst
Comment (dbkinder) [inline doc/kernel/usermode/syscalls.rst]: change to "If any check fails, a kernel oops will be triggered that will kill the"

File: doc/kernel/usermode/syscalls.rst
Comment (dbkinder) [inline doc/kernel/usermode/syscalls.rst]: Change the comma to a period and start a new sentence.

File: doc/porting/arch.rst
Comment (dbkinder) [inline doc/porting/arch.rst]: add a comma after reasons

File: doc/porting/arch.rst
Comment (dbkinder) [inline doc/porting/arch.rst]: should "enable some kind" be "provide some kind" ?

File: doc/porting/arch.rst
Comment (dbkinder) [inline doc/porting/arch.rst]: How about:

     There are no C APIs that need to be implemented to support stack protection,
     and it's entirely implemented within the ``arch/`` code.

File: doc/porting/arch.rst
Comment (dbkinder) [inline doc/porting/arch.rst]: use double back-ticks for filenames:

     ``arch/cpu.h``

File: doc/porting/arch.rst
Comment (dbkinder) [inline doc/porting/arch.rst]: change "this" to "stack protection"

File: doc/porting/arch.rst
Comment (dbkinder) [inline doc/porting/arch.rst]: how do you mark it inaccessible?

File: doc/porting/arch.rst
Comment (dbkinder) [inline doc/porting/arch.rst]: can you add the reference tags to the docs for these macros (as you've done for others)

**Review by dbkinder — CHANGES_REQUESTED** [review]
suggested edits (thanks for creating this!)

File: doc/kernel/threads/lifecycle.rst
Comment (andrewboie) [reply doc/kernel/threads/lifecycle.rst]: I rather strongly disagree here.
This leads to confusion because lower integral priority values have better priority in Zephyr. Priority -1 has "higher" priority than 1.
It is much less confusing to use "better" or "worse" than use "higher" and "lower" since the semantics are flipped

File: doc/kernel/threads/lifecycle.rst
Comment (andrewboie) [reply doc/kernel/threads/lifecycle.rst]: de-initialized actually means something specific and does not imply reset/zero. I can make this clearer in the documentation with a section on initialization states.

File: doc/kernel/usermode/kernelobjects.rst
Comment (andrewboie) [reply doc/kernel/usermode/kernelobjects.rst]: I have no idea, CMake hasn't landed yet

File: doc/kernel/usermode/kernelobjects.rst
Comment (andrewboie) [reply doc/kernel/usermode/kernelobjects.rst]: I could use the term "globals" I guess.
I mean variables that are defined outside of functions

**Comment by andrewboie** [comment]
@dbkinder thanks for reviewing, revising now

File: doc/kernel/threads/lifecycle.rst
Comment (pfalcon) [reply doc/kernel/threads/lifecycle.rst]: Well, and "worse" sounds as if there's something wrong with it (the priority) ;-).

File: doc/kernel/threads/lifecycle.rst
Comment (andrewboie) [reply doc/kernel/threads/lifecycle.rst]: This is defined in the prior "Thread Options" section in this file (line 134-ish)

File: doc/kernel/usermode/syscalls.rst
Comment (andrewboie) [reply doc/kernel/usermode/syscalls.rst]: typically "include" is the verb and "header" is the noun, you can also include C files although this is done rarely. I can still change if you think it's unclear.

File: doc/porting/arch.rst
Comment (andrewboie) [reply doc/porting/arch.rst]: unfortunately this is a set of 10 macros, not just an individual one. how can I refer to a family of macros?

Comment (dbkinder) [reply doc/porting/arch.rst]: nevermind :)

File: doc/kernel/usermode/syscalls.rst
Comment (dbkinder) [reply doc/kernel/usermode/syscalls.rst]: I was reacting to seeing the word "header" four times in two lines, and seeing the phrase "header at the very bottom" (and my brain thinking headers are at the top).  If you're OK, I'm ok.

File: doc/kernel/threads/lifecycle.rst
Comment (dbkinder) [reply doc/kernel/threads/lifecycle.rst]: oops, missed that... I searched for "essential" and it didn't show up (because there were no changes to that section and it was elided by github)

File: doc/kernel/threads/lifecycle.rst
Comment (dbkinder) [reply doc/kernel/threads/lifecycle.rst]: I did a quick grep through the docs for "priority" and all the other Zephyr discussions of threads use "higher" and "lower" (e.g., http://docs.zephyrproject.org/kernel/threads/scheduling.html ).   Also, header files such as kernel.h  talk about "thread of the same or higher priority".

Comment (andrewboie) [reply doc/kernel/threads/lifecycle.rst]: @dbkinder in my opinion the rest of the documentation should be changed to use "better" or "worse" but for the scope of this patch I'll be consistent with the rest of it.

**Comment by andrewboie** [comment]
updated to address @dbkinder feedback, and update to reflect a few minor policy changes introduced by #4754 and #4719

**Review by dbkinder — APPROVED** [review]
thanks

File: doc/kernel/usermode/kernelobjects.rst
Comment (inakypg) [inline doc/kernel/usermode/kernelobjects.rst]: Is there a way to catch this during build time? is it worth it?

Comment (andrewboie) [reply doc/kernel/usermode/kernelobjects.rst]: yes, build time will catch if the kernel object is part of a union and not add it to the gperf table.

File: doc/porting/arch.rst
Comment (inakypg) [reply doc/porting/arch.rst]: you might be able to drop an anchor on the first one and treat it like a group, with a link to said anchor

File: doc/kernel/threads/lifecycle.rst
Comment (mbolivar) [inline doc/kernel/threads/lifecycle.rst]: "uninitialized" typo

File: doc/kernel/threads/lifecycle.rst
Comment (mbolivar) [inline doc/kernel/threads/lifecycle.rst]: I think this is probably cleared up later, but how is it enforced that the stack memory is "unused", and when does memory become "used"?

File: doc/kernel/usermode/kernelobjects.rst
Comment (mbolivar) [inline doc/kernel/usermode/kernelobjects.rst]: "accessible" typo

File: doc/kernel/usermode/kernelobjects.rst
Comment (mbolivar) [inline doc/kernel/usermode/kernelobjects.rst]: I'm not really clear on how a thread object is a permission; what am I missing? Does this mean it can access any objects accessible by the parent thread, except for that parent thread object itself?

File: doc/kernel/usermode/kernelobjects.rst
Comment (mbolivar) [inline doc/kernel/usermode/kernelobjects.rst]: "explicitly" and "implicitly" typos

File: doc/kernel/usermode/kernelobjects.rst
Comment (mbolivar) [inline doc/kernel/usermode/kernelobjects.rst]: Just nitpickery, but I think this should be "permission on *that* object" (at least, I hope so!)

File: doc/kernel/usermode/kernelobjects.rst
Comment (mbolivar) [inline doc/kernel/usermode/kernelobjects.rst]: Presumably it's only safe to use this API when the thread whose access is being revoked can't be scheduled in? Is there an atomicity mechanism other than `k_thread_suspend()` / `k_thread_resume()`?

File: doc/kernel/usermode/kernelobjects.rst
Comment (mbolivar) [inline doc/kernel/usermode/kernelobjects.rst]: "uninitialized"

File: doc/kernel/usermode/kernelobjects.rst
Comment (mbolivar) [inline doc/kernel/usermode/kernelobjects.rst]: "initialized"

File: doc/kernel/usermode/syscalls.rst
Comment (mbolivar) [inline doc/kernel/usermode/syscalls.rst]: "the function"

**Review by mbolivar — COMMENTED** [review]
These are excellent.

I've got minor issues that I don't think should block the merge.

File: doc/kernel/threads/lifecycle.rst
Comment (andrewboie) [reply doc/kernel/threads/lifecycle.rst]: stacks are tracked like other kernel objects, so initialization state is maintained in the associated struct _k_object

File: doc/kernel/usermode/kernelobjects.rst
Comment (andrewboie) [reply doc/kernel/usermode/kernelobjects.rst]: a struct k_thread is itself a kernel object. the child will not have permission inherited to make API calls on its own parent; if this is really desired permission must be granted explicitly.

File: doc/kernel/usermode/kernelobjects.rst
Comment (andrewboie) [reply doc/kernel/usermode/kernelobjects.rst]: it's not atomic, but the permission is checked before the system call implementation is invoked. if another thread revokes permission while a thread is already inside a system call, it is too late.

having said this, i'm spending some time lately making sure there aren't concurrency control issues in the current mechanism wrt permissions

File: doc/kernel/usermode/kernelobjects.rst
Comment (andrewboie) [reply doc/kernel/usermode/kernelobjects.rst]: i'm terrible at typing this word for some reason :)

**Pull request merged**
